New ralter -Wforce option

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
This PR adds support for new '-Wforce' for pbs_ralter command. It will allow admins to force a change to start/end/duration of a reservation even if it is denied by the scheduler.

#### Describe Your Change
In this change, PBS server will enforce the change to the reservation which is being altered with '-Wforce' option. It will do so only when  - 
1 - It is not able to contact the scheduler because scheduling is turned off.
2 - When the scheduler replies with a 'failure' to honor 'ralter' request.


#### Link to Design Doc
[Design ](https://openpbs.atlassian.net/wiki/spaces/PD/pages/1996947457/New+-Wforce+option+for+pbs+ralter+to+enforce+changes+to+start+end+duration+of+a+reservation.)
[Forum discussion](https://community.openpbs.org/t/pbs-ralter-wforce-option/2228)


#### Attach Test and Valgrind Logs/Output
[test_multisched.txt](https://github.com/openpbs/openpbs/files/5059052/test_multisched.txt)
[test_resv.txt](https://github.com/openpbs/openpbs/files/5059053/test_resv.txt)
[test_ralter.txt](https://github.com/openpbs/openpbs/files/5059055/test_ralter.txt)

bhroam commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -2595,3 +2615,141 @@ def test_alter_duration_standing_resv_future_occrs(self):
+        # Ideally, in 1 hr second occurrence of reservation will start running
+        # and it will run for 3 mins. This means the new reservation will be
+        # confirmed.
+        new_offset = (start + 3800) - time.time()
```
You need to int() time.time()

arungrover commented on src/cmds/pbs_ralter.c:
```diff
@@ -153,13 +156,31 @@ process_opts(int argc, char **argv, struct attrl **attrp, char *dest)
+					fprintf(stderr, "pbs_ralter: illegal -W value\n");
+					errflg++;
+					break;
+				}
```
oh this is for the case where -W is given with anything other than 'force'. Like -Wextend etc.

bhroam replied:
If you say -Wextend, then optarg will be "extend", and strlen() will be 6.  The only way strlen() will be 0 is if optarg is not given, and if that is the case, getopt() will complain.

arungrover commented on src/cmds/pbs_ralter.c:
```diff
@@ -51,6 +51,7 @@
 static struct attrl *attrib = NULL;
 static time_t dtstart;
 static time_t dtend;
+int force_alter = FALSE;
```
I can make it a static, but I do want to initialize explicitly because if TRUE or FALSE ever change their value (highly unlikely) then it may not work.
If you feel strongly about this change then let me know, I'll make it.

bhroam replied:
Initializing it is fine.  I personally don't think any of those need to be static since this is a command, and we don't link them into other programs.  I was just thinking it should be consistent with what is there.

arungrover commented on src/server/req_modify.c:
```diff
@@ -1078,6 +1153,40 @@ req_modifyReservation(struct batch_request *preq)
+					resv_revert_alter(presv);
+					return;
+				}
+				append_link(&presv->ri_svrtask, &pwt->wt_linkobj, pwt);
```
Accounting record is one reason, another reason is that if I don't create a local request then I'll have to replicate a lot of code from req_confirmresv to alter existing timed events. I thought this was a cleaner approach.

arungrover commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -2243,3 +2243,107 @@ def test_resv_from_job_in_multi_sched_using_rsub(self):
+            self.server.expect(RESV, attr, rid)
+            partition = self.server.status(RESV, 'partition', id=rid)
+            if (partition[0]['partition'] == 'P1'):
+                p1_start_time = a['reserve_start']
```
Even if I do that, I cannot avoid status call because I need to get the reservation that's confirmed in partition P1. So while I have the status, it wouldn't be too bad to get it from the reservation :)

bhroam replied:
Ahh OK, so you need to keep the status().  I'd still keep the reservation start and just use it.

arungrover commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -2243,3 +2243,107 @@ def test_resv_from_job_in_multi_sched_using_rsub(self):
+        # reservation confimed in partition P1
+        bu = BatchUtils()
+        stime = int(time.time()) + 30
+        etime = bu.convert_stime_to_seconds(p1_start_time) - 10
```
You are right, this is a compelling reason to make the change.

arungrover commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -2243,3 +2243,107 @@ def test_resv_from_job_in_multi_sched_using_rsub(self):
+        etime = etime + 300
+        attrs['reserve_end'] = bu.convert_seconds_to_datetime(etime)
+        attrs['reserve_start'] = bu.convert_seconds_to_datetime(stime)
+        del attrs['Resource_List.select']
```
I think It does not matter if I reuse the dict. Please let me know if you feel strongly about creating a new dict.

bhroam replied:
I'd say recreate it.  You have to set/delete 3 things this way, where you only have to set 2 things if you recreate it.

arungrover commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -2243,3 +2243,107 @@ def test_resv_from_job_in_multi_sched_using_rsub(self):
+        Test that in a multi-sched setup when all schedulers are disabled
+        ralter -Wforce can still modify a reservation successfully even when
+        the ralter results into over subscription of resources.
+        """
```
It does not actually, I think this test will exercise the same code as the one written in ralter test suite for unconfirmed reservation. I'll remove this test. Thanks for catching this.

vstumpf replied:
has this been addressed?

arungrover replied:
I did remove that test but the docstring of the second test was wrong. I'll change that

arungrover commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -2595,3 +2615,141 @@ def test_alter_duration_standing_resv_future_occrs(self):
+        attrs['reserve_end'] = new_end_conv
+
+        self.server.alterresv(rid, attrs, extend='force')
+        msg = "pbs_ralter: " + rid + " CONFIRMED"
```
I know :)

arungrover commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -2595,3 +2615,141 @@ def test_alter_duration_standing_resv_future_occrs(self):
+        new_end_conv = self.bu.convert_seconds_to_datetime(new_end)
+        attrs['reserve_end'] = new_end_conv
+
+        self.server.alterresv(rid, attrs, extend='force')
```
It was getting too complex to change alter_a_reservation to check for confirmed ralter for an unconfirmed reservation. I thought I am better off just calling alter myself.

arungrover commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -2595,3 +2615,141 @@ def test_alter_duration_standing_resv_future_occrs(self):
+        new_end = end - 3600
+        new_duration = duration - 1800
+        new_start_conv = self.bu.convert_seconds_to_datetime(new_start)
+        attrs['reserve_start'] = new_start_conv
```
oh, since I'm not calling alter_a_reservation, this conversion is needed

bhroam replied:
alterresv() doesn't do this conversion for you?  That's bad.  That makes it different than submit() which does.

arungrover commented:
@bhroam thanks for reviewing the code. I've addressed your comments. Please have a look.
[test_multisched.txt](https://github.com/openpbs/openpbs/files/5066463/test_multisched.txt)
[test_resv.txt](https://github.com/openpbs/openpbs/files/5066465/test_resv.txt)
[test_ralter.txt](https://github.com/openpbs/openpbs/files/5066468/test_ralter.txt)

bhroam commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -130,6 +130,11 @@ def get_tzid(self):
             tzone = 'America/Los_Angeles'
         return tzone
 
+    def set_scheduling(self, scheds=[], op=False):
```
Don't default to [].  This will create a single list which will be used for all function calls.  You don't modify it, but if you had, the changes would persist from call to call.  If you want an empty list, you set the default to None, and then check for it and set scheds to [] in the function itself.

An option is to set scheds to self.server.scheds.  That way the default is all the schedulers.

vstumpf commented on src/server/req_rescq.c:
```diff
@@ -589,12 +590,14 @@ req_confirmresv(struct batch_request *preq)
-					presv->ri_brp = NULL;
+					if (!(presv->ri_alter.ra_flags & RESV_ALTER_FORCED)) {
+						presv->ri_wattr[RESV_ATR_interactive].at_flags &= ~ATR_VFLAG_SET;
+						snprintf(buf, sizeof(buf), "%s DENIED",
```
Since you're here, can you put these on one line? instead of putting them on multiple.

vstumpf commented on src/server/svr_chk_owner.c:
```diff
@@ -575,6 +575,10 @@ svr_authorize_resvReq(struct batch_request *preq, resc_resv *presv)
 		return (0);
+	/* Only Manager has privilage to force modify reservation */
+	if (preq->rq_type == PBS_BATCH_ModifyResv && (preq->rq_extend != NULL) &&
+	    (strcmp(preq->rq_extend, FORCE) == 0) && ((preq->rq_perm & ATR_DFLAG_MGWR) == 0))
```
Should this be strstr instead of strcmp? What if we add more things to ralter -W, and "Force" isn't the first one?

arungrover replied:
You are right, but we don't know what will be the next option name (maybe the new name has 'force' in it too) and if multiple options can be given within one extend field then I think we will have to break it on a delimiter and then do strcmp.
I guess this is fine for now since this is the only option. Please let me know if you think otherwise.

vstumpf replied:
Your explanation works for me

vstumpf commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -317,8 +325,13 @@ def alter_a_reservation(self, r, start, end, shift=0,
-        :type a_duration: int
+        :param a_duration: The duration to modify.
+        :type a_duration: int.
+        :param force: Force ralter.
```
docstring says force but the function declaration says extend

arungrover replied:
Good catch!

arungrover commented:
Thanks @vstumpf  and @bhroam for reviewing the changes. I've addressed your comments. I also added another test about using ralter -Wforce on a running reservation.

arungrover commented:
@vstumpf, @bhroam  Gentle reminder.
Thanks!

bhroam commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -2753,3 +2753,45 @@ def test_alter_force_select(self):
+        self.alter_a_reservation(rid, start, end, confirm=False, shift=-100,
+                                 alter_e=True, extend='force')
+        _, _, t_end = self.get_resv_time_info(rid)
+        end = end - 100
```
Consider using -=

bhroam commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -2753,3 +2753,45 @@ def test_alter_force_select(self):
+        self.assertEqual(int(t_end), end)
+
+        self.alter_a_reservation(rid, start, end, confirm=False,
+                                 a_duration=4000, extend='force', sequence=2)
```
Usually you have to bump sequence for each alter you do.  This is the 3rd alter, why didn't you need  to bump it in the last call?

arungrover replied:
Actually first alter command is rejected because there is a running job inside this reservation. This is the reason the sequence number here is 2.

arungrover replied:
I've added a comment to make it clear.
